File: apps/mobile/ios/ParentAssist/ViewModels/SessionViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class SessionViewModel: ObservableObject {
    enum State: Equatable {
        case idle
        case preparing
        case sharing(ShareSession)
        case error(String)
    }

    @Published private(set) var state: State = .idle
    @Published var hasAcceptedConsent = false

    private let sessionService: SessionServicing
    private let screenShareController: ScreenShareControlling
    private let webrtcStreamer: WebRTCStreaming

    init(
        sessionService: SessionServicing,
        screenShareController: ScreenShareControlling,
        webrtcStreamer: WebRTCStreaming
    ) {
        self.sessionService = sessionService
        self.screenShareController = screenShareController
        self.webrtcStreamer = webrtcStreamer
    }

    func startSession() {
        guard hasAcceptedConsent else {
            state = .error("Please review and accept the sharing consent.")
            return
        }

        state = .preparing
        Task {
            do {
                let session = try await sessionService.createSession()
                try await webrtcStreamer.connect(sessionId: session.sessionId)
                try await screenShareController.startSharing()
                state = .sharing(session)
            } catch {
                state = .error(error.localizedDescription)
            }
        }
    }

    func stopSession() {
        guard case let .sharing(session) = state else { return }

        Task {
            await webrtcStreamer.disconnect()
            await screenShareController.stopSharing()
            await sessionService.stopSession(session)
            state = .idle
        }
    }

    func resetError() {
        if case .error = state {
            state = .idle
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/example/parentassist/SessionViewModel.kt
package com.example.parentassist

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

sealed class SessionState {
    data object Idle : SessionState()
    data object Preparing : SessionState()
    data class Sharing(val session: ShareSession) : SessionState()
    data class Error(val message: String) : SessionState()
}

class SessionViewModel(
    private val sessionService: SessionService,
    private val screenShareController: ScreenShareController
) : ViewModel() {
    private val _state = MutableStateFlow<SessionState>(SessionState.Idle)
    val state: StateFlow<SessionState> = _state

    var hasAcceptedConsent: Boolean = false
        private set

    fun updateConsent(accepted: Boolean) {
        hasAcceptedConsent = accepted
        if (_state.value is SessionState.Error) {
            _state.value = SessionState.Idle
        }
    }

    fun startSession() {
        if (!hasAcceptedConsent) {
            _state.value = SessionState.Error("Please review and accept the sharing consent.")
            return
        }

        _state.value = SessionState.Preparing
        viewModelScope.launch {
            try {
                val session = sessionService.createSession()
                screenShareController.startSharing()
                _state.value = SessionState.Sharing(session)
            } catch (exception: Exception) {
                _state.value = SessionState.Error(exception.message ?: "Unable to start sharing.")
            }
        }
    }

    fun stopSession() {
        val current = _state.value
        if (current is SessionState.Sharing) {
            viewModelScope.launch {
                screenShareController.stopSharing()
                sessionService.stopSession(current.session)
                _state.value = SessionState.Idle
            }
        }
    }
}

class SessionViewModelFactory(
    private val sessionService: SessionService,
    private val screenShareController: ScreenShareController
) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SessionViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return SessionViewModel(sessionService, screenShareController) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}
